### PR TITLE
Relative erosion

### PR DESCRIPTION
**Pull request by PoyiOstona**
Have a look at this option of having different output for erosion models. Tested for bend model. If you like it, we can develop erosion models in this way and the user will have an option of choosing output format.
P.S.. Third option, Actual surface thickness loss (DNVGL eq. 4.36) can also be added easily.

### CODE REVIEW

**Comment by PoyiOstona** [comment]
Unit tests related to calculation assertion shall be updated too.

File: pysand/erosion.py [hunk 1]
```diff
@@ -246,52 +244,55 @@ def welded_joint(v_m, rho_m, Q_s, D, d_p, h, alpha=60, material='duplex'):
...
         C2 = 1
-    E_up = K * F(a_rad, ad) * v_m**n * np.sin(a_rad) / (rho_t * A_pipe) * C2 * C_unit * Q_s / 1000
-    E_down = 3.3e-2 * (7.5e-4 + h) * v_m**n * D**(-2) * Q_s / 1000
-    return E_up, E_down
+    if location == 'downstream':
+        E_down = 3.3e-2 * (7.5e-4 + h) * v_m**n * D**(-2) * (1e6/C_unit)
+        return E_down
+    elif location == 'upstream':
+        E_up = K * F(a_rad, ad) * v_m ** n * np.sin(a_rad) / (rho_t * A_pipe) * C2 * 10**6
+        return E_up
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply pysand/erosion.py]: Is error only on this one? Or does any other erosion models contain error? I mean, i f you have QCed them against Excel sheet.

[the remark this answers is not in the document]
Comment (knudsvik) [reply pysand/erosion.py]: I only found error on this one model, I checked all models - the others were fine!

File: pysand/erosion.py [hunk 2]
```diff
@@ -246,52 +244,55 @@ def welded_joint(v_m, rho_m, Q_s, D, d_p, h, alpha=60, material='duplex'):
...
     if validate_inputs(**kwargs):
         return np.nan
 
     R = Dm / D - 0.5  # Synthetic bend radius
-    return bend(v_m, rho_m, mu_m, Q_s, R, GF, D, d_p, material=material)
+    return bend(v_m, rho_m, mu_m, R, GF, D, d_p, material=material)  # Relative surface thickness loss [mm/t]
 
 
-def reducer(v_m, rho_m, Q_s, D1, D2, d_p, GF=2, alpha=60, material='duplex'):
+def reducer(v_m, rho_m, D1, D2, d_p, GF=2, alpha=60, material='duplex'):
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply pysand/erosion.py]: Yes, agree with this notation. Also, I see alpha for reducer model (page 33) should be between 10 and 80 deg. But in the validation we have common 10 to 90 deg, I guess this one is valid for probes.

[the remark this answers is not in the document]
Comment (knudsvik) [reply pysand/erosion.py]: Yes, I see that now. It should be validated for 10 to 80 and give a warning outside probably

File: pysand/erosion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (knudsvik) [reply pysand/erosion.py]: As discussed, this was not an error ;)

File: tests/test_erosion.py [hunk 3]
```diff
@@ -272,58 +289,75 @@ def test_return_nan():
...
-        kwargs[inp] = -1
-        assert np.isnan(straight_pipe(**kwargs))
+    kwargs = {'v_m': -1, 'D': D}
+    assert np.isnan(straight_pipe(**kwargs))
 
     # test welded joint
-    kwargs = {'v_m': v_m, 'rho_m': rho_m, 'Q_s': Q_s, 'D': D, 'd_p': d_p, 'h': h}
-    for inp in ['v_m', 'rho_m', 'Q_s']:
+    kwargs = {'v_m': v_m, 'rho_m': rho_m, 'D': D, 'd_p': d_p, 'h': h}
+    for inp in ['v_m', 'rho_m']:
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply tests/test_erosion.py]: This test is for throwing np.nan. Default alpha is used.

File: tests/test_erosion.py [hunk 4]
```diff
@@ -272,58 +289,75 @@ def test_return_nan():
...
         assert np.isnan(manifold(**kwargs))
+        kwargs['v_m'] = v_m
+        kwargs['rho_m'] = rho_m
+        kwargs['mu_m'] = mu_m
 
     # test reducer
-    kwargs = {'v_m': v_m, 'rho_m': rho_m, 'Q_s': Q_s, 'D1': D1, 'D2': D2, 'd_p': d_p}
-    for inp in ['v_m', 'rho_m', 'Q_s']:
+    kwargs = {'v_m': v_m, 'rho_m': rho_m, 'D1': D1, 'D2': D2, 'd_p': d_p}
+    for inp in ['v_m', 'rho_m']:
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply tests/test_erosion.py]: This test is for throwing np.nan. Default alpha is used.

File: tests/test_erosion.py [hunk 5]
```diff
@@ -272,58 +289,75 @@ def test_return_nan():
...
         assert np.isnan(reducer(**kwargs))
+        kwargs['v_m'] = v_m
+        kwargs['rho_m'] = rho_m
 
     # test probes
-    # v_m, rho_m, Q_s, D, d_p
-    kwargs = {'v_m': v_m, 'rho_m': rho_m, 'Q_s': Q_s, 'D': D, 'd_p': d_p}
-    for inp in ['v_m', 'rho_m', 'Q_s']:
+    kwargs = {'v_m': v_m, 'rho_m': rho_m, 'D': D, 'd_p': d_p}
+    for inp in ['v_m', 'rho_m']:
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply tests/test_erosion.py]: This test is for throwing np.nan. Default alpha is used.

File: pysand/erosion.py [hunk 6]
```diff
@@ -35,8 +34,8 @@ def validate_inputs(**kwargs):
 
     for i in ['v_m', 'rho_m', 'mu_m', 'Q_s']:
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply pysand/erosion.py]: Added alpha into list of static inputs - line 59

File: tests/test_erosion.py [hunk 7]
```diff
@@ -4,17 +4,19 @@
...
 from pysand.erosion import validate_inputs, bend, tee, straight_pipe, \
-    welded_joint, manifold, reducer, probes, flexible, choke_gallery, F, material_properties
+    welded_joint, manifold, reducer, probes, flexible, choke_gallery, F, material_properties, erosion_rate
 
 def test_validate_inputs(caplog):
 
+    num = 1
     # Testing input throws exception throws exception
-    kwargs = {'v_m': 29.3, 'rho_m': 30, 'mu_m': 1.5e-5, 'Q_s': 2400*1000/86400/365}
+    kwargs = {'v_m': num, 'rho_m': num, 'mu_m': num, 'Q_s': num}
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply tests/test_erosion.py]: Added alpha into list of static inputs - line 59

File: pysand/erosion.py [hunk 8]
```diff
@@ -78,8 +77,8 @@ def validate_inputs(**kwargs):
         if kwargs['GF'] not in [1, 2, 3, 4]:
             logger.warning('Geometry factor, GF, can only be 1, 2, 3 or 4')
     if 'alpha' in kwargs:
-        if (kwargs['alpha'] < 10) or (kwargs['alpha'] > 90):
-            logger.warning('Particle impact angle [degrees], alpha, is outside RP-O501 model boundaries (10-90 deg).')
+        if (kwargs['alpha'] < 10) or (kwargs['alpha'] > 80):
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply pysand/erosion.py]: I also think we should then separate alpha validation within each models

File: tests/test_erosion.py [hunk 9]
```diff
@@ -4,17 +4,19 @@
...
     for inp in ['v_m', 'rho_m', 'mu_m', 'Q_s']:
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply tests/test_erosion.py]: Added alpha into list of static inputs - line 59

File: tests/test_erosion.py [hunk 10]
```diff
@@ -98,7 +106,7 @@ def test_validate_inputs(caplog):
             validate_inputs(**kwargs)
             info = [record for record in caplog.records if record.levelname == 'WARNING']
             assert any(
-                "Particle impact angle [degrees], alpha, is outside RP-O501 model boundaries (10-90 deg)."
+                "Particle impact angle [degrees], alpha, is outside RP-O501 model boundaries (10-80 deg)."
```
[the remark this answers is not in the document]
Comment (PoyiOstona) [reply tests/test_erosion.py]: Added separate alpha validations for each model

**Pull request merged**